hedera-node/src/main/java/com/hedera/services/store/tokens/HederaTokenStore.java

```diff
@@ -225,52 +228,75 @@ public void setHederaLedger(final HederaLedger hederaLedger) {
 	@Override
 	public ResponseCodeEnum associate(AccountID aId, List<TokenID> tokens, boolean automaticAssociation) {
 		return fullySanityChecked(true, aId, tokens, (account, tokenIds) -> {
-			final var accountTokens = (MerkleAccountTokens) accountsLedger.get(aId, TOKENS);
+			final var lastAssociatedToken = (EntityNumPair) accountsLedger.get(aId, LAST_ASSOCIATED_TOKEN);
+			final List<TokenID> listOfAssociatedTokens = new ArrayList<>();
+			var currKey = new EntityNumPair(lastAssociatedToken.value());
+			// if this lastAssociatedToken == 0 then this is account has No token associations currently
+			if (!lastAssociatedToken.equals(MISSING_NUM_PAIR)) {
+				// get All the tokenIds associated by traversing the linkedList
+				while (!currKey.equals(MISSING_NUM_PAIR)) {
+					listOfAssociatedTokens.add(currKey.asAccountTokenRel().getRight());
+					currKey = (EntityNumPair) tokenRelsLedger.get(currKey.asAccountTokenRel(), NEXT_KEY);
+				}
+			}
+
 			for (var id : tokenIds) {
-				if (accountTokens.includes(id)) {
+				if (listOfAssociatedTokens.contains(id)) {
```

Comment:
This check should be done via a `MerkleMap.containsKey()` on the `tokenAssociations` map---no need to explicitly construct the list of `TokenID`s.